File: app/src/androidTest/java/network/minter/bipwallet/tests/ui/SendCoinsTabTest.java
/*
 * Copyright (C) by MinterTeam. 2019
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.tests.ui;

import android.support.test.espresso.IdlingRegistry;
import android.support.test.espresso.ViewInteraction;
import android.support.test.espresso.intent.Intents;
import android.support.test.espresso.matcher.ViewMatchers;
import android.support.test.filters.LargeTest;
import android.support.test.rule.ActivityTestRule;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.math.BigDecimal;

import network.minter.bipwallet.R;
import network.minter.bipwallet.advanced.models.SecretData;
import network.minter.bipwallet.advanced.repo.SecretStorage;
import network.minter.bipwallet.home.ui.HomeActivity;
import network.minter.bipwallet.internal.auth.AuthSession;
import network.minter.bipwallet.internal.system.testing.CallbackIdlingResource;
import network.minter.bipwallet.sending.ui.SendTabFragment;
import network.minter.bipwallet.settings.repo.MinterBotRepository;
import network.minter.bipwallet.tests.internal.TestWallet;
import network.minter.blockchain.models.operational.OperationType;
import network.minter.core.MinterSDK;
import network.minter.core.bip39.MnemonicResult;
import network.minter.core.crypto.MinterAddress;
import network.minter.profile.models.User;
import timber.log.Timber;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.action.ViewActions.replaceText;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDescendantOfA;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.isEnabled;
import static android.support.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withSubstring;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static network.minter.bipwallet.internal.helpers.MathHelper.bdHuman;
import static network.minter.bipwallet.tests.internal.MyMatchers.withInputLayoutError;
import static network.minter.bipwallet.tests.internal.MyMatchers.withInputLayoutHint;
import static org.hamcrest.CoreMatchers.allOf;
import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.Matchers.not;

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
@LargeTest
@RunWith(MockitoJUnitRunner.class)
public class SendCoinsTabTest extends BaseUiTest {

    @Rule
    public ActivityTestRule<HomeActivity> mActivityTestRule = new ActivityTestRule<>(HomeActivity.class, true, false);
    private MinterAddress mAddress;

    @Before
    public void setUp() {
        super.setUp();
        final SecretStorage ss = TestWallet.app().secretStorage();
        ss.destroy();

        MnemonicResult mnemonicResult = generateMnemonic();
        mAddress = ss.add(mnemonicResult);

        TestWallet.app().session().login(
                AuthSession.AUTH_TOKEN_ADVANCED,
                new User(AuthSession.AUTH_TOKEN_ADVANCED),
                AuthSession.AuthType.Advanced
        );

        MinterBotRepository botRepo = new MinterBotRepository();
        MinterBotRepository.MinterBotResult botResult = botRepo.requestFreeCoins(mAddress).blockingFirst();
        if (!botResult.isOk()) {
            throw new RuntimeException("Unable to get free coins, which is required for testing!");
        }
        TestWallet.app().accountStorageCache().update(true);
        mActivityTestRule.launchActivity(null);

        Timber.d("Secret storage by address %s: %s", mAddress, TestWallet.app().secretStorage().getSecret(mAddress) != null ? "true" : "false");
        Intents.init();
    }

    @After
    public void tearDown() {
        super.tearDown();
        Intents.release();
        mActivityTestRule.finishActivity();
    }

    @Test
    public void testSendCoinsInsufficient() {
        selectTab(1);

        // wait while balance be a 100
        waitForBalance(100);
        // wait while balance isn't update
        waitForBalanceUpdate();


        BigDecimal amountToSend = new BigDecimal(10000);

        // register idlings
        SendTabFragment fragment = ((SendTabFragment) mActivityTestRule.getActivity().getActiveTabs().get(1));
        CallbackIdlingResource confirmIdling = new CallbackIdlingResource();
        CallbackIdlingResource completeIdling = new CallbackIdlingResource();

        IdlingRegistry.getInstance().register(confirmIdling);
        IdlingRegistry.getInstance().register(completeIdling);

        fragment.registerIdlings(confirmIdling, completeIdling);

        SecretData sd = SecretStorage.generateAddress();
        final String address = sd.getMinterAddress().toString();

        // recipient
        ViewInteraction recipient = onView(withId(R.id.input_recipient));
        recipient.perform(replaceText(address));
        recipient.check(matches(withText(address)));

        // amount
        ViewInteraction amount = onView(withId(R.id.input_amount));
        amount.perform(replaceText(amountToSend.toString()));

        // submit
        ViewInteraction submit = onView(allOf(withId(R.id.action), withText(R.string.btn_send)));
        submit.check(matches((isEnabled())));

        submit.perform(click());


        ViewInteraction confirmTitle = onView(allOf(isDescendantOfA(withId(android.R.id.content)), withId(R.id.title)));
        confirmTitle.check(matches(withText("You're sending")));

        onView(withId(R.id.dialog_amount)).check(matches(withText(String.format("%s %s", bdHuman(amountToSend), MinterSDK.DEFAULT_COIN))));
        onView(withId(R.id.tx_recipient_name)).check(matches(withText(address)));


        ViewInteraction confirm = onView(allOf(withId(R.id.action_confirm), withText(R.string.btn_send)));
        confirm.perform(click());

        onView(allOf(isDescendantOfA(withId(android.R.id.content)), withId(R.id.title)))
                .check(matches(
                        withText("Unable to send transaction"))
                );

        onView(allOf(isDescendantOfA(withId(android.R.id.content)), withId(R.id.dialog_text)))
                .check(matches(
                        allOf(
                                withText(containsString("Insufficient")),
                                withText(containsString(mAddress.toString()))
                        )
                ));

        // close dialog to prevent leaking
        onView(allOf(withText("Close"), withId(R.id.action_confirm))).perform(click());


        IdlingRegistry.getInstance().unregister(confirmIdling);
        IdlingRegistry.getInstance().unregister(completeIdling);

        // as before
        waitForBalance(100);
    }

    @Test
    public void testSendCoinsSuccessfully() {
        selectTab(1);

        waitForBalance(100d);
        waitForBalanceUpdate();

        // register idlings
        SendTabFragment fragment = ((SendTabFragment) mActivityTestRule.getActivity().getActiveTabs().get(1));
        CallbackIdlingResource confirmIdling = new CallbackIdlingResource();
        CallbackIdlingResource completeIdling = new CallbackIdlingResource();

        IdlingRegistry.getInstance().register(confirmIdling);
        IdlingRegistry.getInstance().register(completeIdling);

        fragment.registerIdlings(confirmIdling, completeIdling);

        SecretData sd = SecretStorage.generateAddress();
        final String address = sd.getMinterAddress().toString();

        // recipient
        ViewInteraction recipient = onView(withId(R.id.input_recipient));
        recipient.perform(replaceText(address));
        recipient.check(matches(withText(address)));

        // amount
        ViewInteraction amount = onView(withId(R.id.input_amount));
        amount.perform(replaceText("1"));

        // submit
        ViewInteraction submit = onView(allOf(withId(R.id.action), withText(R.string.btn_send)));
        submit.check(matches((isEnabled())));

        submit.perform(click());


        ViewInteraction confirmTitle = onView(allOf(isDescendantOfA(withId(android.R.id.content)), withId(R.id.title)));
        confirmTitle.check(matches(withText("You're sending")));

        onView(withId(R.id.dialog_amount)).check(matches(withText(String.format("%s %s", bdHuman(1d), MinterSDK.DEFAULT_COIN))));
        onView(withId(R.id.tx_recipient_name)).check(matches(withText(address)));


        ViewInteraction confirm = onView(allOf(withId(R.id.action_confirm), withText(R.string.btn_send)));
        confirm.perform(click());

        onView(withId(R.id.tx_description)).check(matches(withText(R.string.tx_send_success_dialog_description)));
        onView(withId(R.id.tx_recipient_name)).check(matches(withText(address)));
        onView(withId(R.id.action_view_tx))
                .check(matches(isDisplayed()))
                .check(matches(withText("VIEW TRANSACTION")));

        // close dialog to prevent leaking
        onView(allOf(withText("Close"), withId(R.id.action_close))).perform(click());

        IdlingRegistry.getInstance().unregister(confirmIdling);
        IdlingRegistry.getInstance().unregister(completeIdling);

        // 100 - 1 - 0.01 fee  - 98.99
        waitForBalance(new BigDecimal(99).subtract(OperationType.SendCoin.getFee()));

    }

    @Test
    public void testAccountSelector() {
        selectTab(1);

        waitForBalance(100d);
        final String balanceString = String.format("%s (%s)", MinterSDK.DEFAULT_COIN, bdHuman(new BigDecimal(100.d)));

        ViewInteraction submit = onView(allOf(withId(R.id.action), withText(R.string.btn_send)));
        submit.check(matches(not(isEnabled())));

        ViewInteraction accountInput = onView(withId(R.id.input_coin));
        accountInput.check(matches(withSubstring(MinterSDK.DEFAULT_COIN)));
        accountInput.check(matches(withText(balanceString)));

        accountInput.perform(click());
        ViewInteraction accountDialogTitle = onView(withText(R.string.title_select_account));
        accountDialogTitle.check(matches(isDisplayed()));
        ViewInteraction accountItem = onView(allOf(withId(R.id.item_title), withText(balanceString)));
        accountItem.perform(click());

        // check submit still not enabled
        submit.check(matches(not(isEnabled())));


        ViewInteraction recipient = onView(withId(R.id.input_recipient));

        recipient.check(matches(not(withText("aaa"))));
        recipient.check(matches(withInputLayoutHint(R.string.tx_send_recipient_hint)));
        // set something strange
        {
            recipient.perform(replaceText("aaa"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set something illegal
        {
            recipient.perform(replaceText("whywhywhye"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set too short username, minimum 5 chars
        {
            recipient.perform(replaceText("@aabb"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set too long username, max 16 chars
        {
            recipient.perform(replaceText("@abcdefghijklmnopqrtstuwxyz"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set invalid email #1
        {
            recipient.perform(replaceText("test@fly"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set invalid email #2
        {
            recipient.perform(replaceText("test.fly.com"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set invalid email #2 (but this will parsed as invalid username)
        {
            recipient.perform(replaceText("@gmail.com"));
            recipient.check(matches(withInputLayoutError("Incorrect recipient format")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }
        // set valid username, minimum 5 chars
        {
            recipient.perform(replaceText("@aabbc"));
            recipient.check(matches(withInputLayoutError(null)));
            // check submit still not enabled
            submit.check(matches((isEnabled())));
        }

        // set valid username, minimum 5 chars
        {
            recipient.perform(replaceText("@12345"));
            recipient.check(matches(withInputLayoutError(null)));
            // check submit still not enabled
            submit.check(matches((isEnabled())));
        }

        ViewInteraction amount = onView(withId(R.id.input_amount));
        amount.check(matches(withText("")));
        amount.check(matches(withInputLayoutHint(R.string.label_amount)));

        // set 0
        {
            amount.perform(replaceText("0"));
            amount.check(matches(withInputLayoutError("Amount must be greater than 0")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set .
        {
            amount.perform(replaceText("."));
            amount.check(matches(withInputLayoutError("Amount must be greater than 0")));
            // check submit still not enabled
            submit.check(matches(not(isEnabled())));
        }

        // set .1
        {
            amount.perform(replaceText(".1"));
            amount.check(matches(withInputLayoutError(null)));
            // check submit is enabled
            submit.check(matches((isEnabled())));
        }

        // maximum is 10^-18
        {
            amount.perform(replaceText("0.102030405060708090"));
            amount.check(matches(withInputLayoutError(null)));
            // check submit is enabled
        }

        ViewInteraction useMaxButton = onView(withId(R.id.action_maximum));
        useMaxButton.check(matches(withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
        useMaxButton.perform(click());

        amount.check(matches(withText("100")));

        ViewInteraction feeLabel = onView(withId(R.id.fee_label));
        feeLabel.check(matches(withText(R.string.tx_send_fee_hint)));
        ViewInteraction feeText = onView(withId(R.id.fee_value));
        feeText.check(matches(withText(
                String.format("%s %s", bdHuman(OperationType.SendCoin.getFee()), MinterSDK.DEFAULT_COIN)
        )));

    }
}

File: app/src/androidTest/java/network/minter/bipwallet/tests/ui/SendCoinsTabTest.kt
/*
 * Copyright (C) by MinterTeam. 2020
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.tests.ui

import network.minter.bipwallet.home.ui.HomeActivity
import network.minter.bipwallet.internal.auth.AuthSession
import network.minter.bipwallet.internal.storage.SecretStorage
import network.minter.bipwallet.tests.LazyActivityScenarioRule
import network.minter.bipwallet.tests.internal.TestWallet
import network.minter.core.crypto.MinterAddress
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import timber.log.Timber

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
class SendCoinsTabTest : BaseUiTest() {
    @get:Rule
    var activityTestRule = LazyActivityScenarioRule(false, HomeActivity::class.java)

    private var mAddress: MinterAddress? = null

    @Before
    override fun setUp() {
        super.setUp()
        val ss: SecretStorage = TestWallet.app().secretStorage()
        ss.destroy()
        val mnemonicResult = generateMnemonic()
        mAddress = ss.add(mnemonicResult)
        TestWallet.app().session().login(
                AuthSession.AUTH_TOKEN_ADVANCED,
                AuthSession.AuthType.Advanced
        )

        TestWallet.app().accountStorageCache().update(true)


        activityTestRule.launch(null)
//        TestWallet.app().idlingManager().register()
    }

    @After
    override fun tearDown() {
        Timber.d("UnRegister idlings!")
//        TestWallet.app().idlingManager().unregister()
        super.tearDown()

    }

    @Test
    @Throws(InterruptedException::class)
    fun testSendCoinCommission() {
//        val gateMock = ApiMockInterceptor("gate", activityTestRule.activity)
//        gateMock.override("api/v1/min-gas", "/v1/min-gas/min_gas_price_1")
//        MinterExplorerApi.getInstance().getGateApiService().addHttpInterceptor(gateMock)
//        selectTab(1)
//        val sd: SecretData = SecretStorage.generateAddress()
//        val address: String = sd.getMinterAddress().toString()
//        val amountToSend = BigDecimal("1")
//
//        // recipient
//        val recipient = Espresso.onView(ViewMatchers.withId(R.id.input_recipient))
//        recipient.perform(ViewActions.replaceText(address))
//        recipient.check(ViewAssertions.matches(ViewMatchers.withText(address)))
//
//        // amount
//        val amount = Espresso.onView(ViewMatchers.withId(R.id.input_amount))
//        amount.perform(ViewActions.replaceText(amountToSend.toString()))
//
//        // submit
//        val submit = Espresso.onView(CoreMatchers.allOf(ViewMatchers.withId(R.id.action), ViewMatchers.withText(R.string.btn_send)))
//        submit.check(ViewAssertions.matches(ViewMatchers.isEnabled()))
//
//        // fee
//        val feeValue = Espresso.onView(ViewMatchers.withId(R.id.fee_value))
//
//
//        // fee = 0.01
//        feeValue.check(ViewAssertions.matches(ViewMatchers.withText(String.format("%s %s", bdHuman(OperationType.SendCoin.fee.multiply(BigDecimal("1"))), MinterSDK.DEFAULT_COIN))
//        ))
//        Thread.sleep(1000 * 3.toLong())
//
//
//        // fee = 0.02
//        gateMock.override("api/v1/min-gas", "/v1/min-gas/min_gas_price_2")
//        amount.perform(ViewActions.replaceText(amountToSend.toString()))
//        feeValue.check(ViewAssertions.matches(ViewMatchers.withText(String.format("%s %s", bdHuman(OperationType.SendCoin.fee.multiply(BigDecimal("2"))), MinterSDK.DEFAULT_COIN))
//        ))
//        Thread.sleep(1000 * 3.toLong())
//
//        // fee = 0.04
//        gateMock.override("api/v1/min-gas", "/v1/min-gas/min_gas_price_4")
//        amount.perform(ViewActions.replaceText(amountToSend.toString()))
//        feeValue.check(ViewAssertions.matches(ViewMatchers.withText(String.format("%s %s", bdHuman(OperationType.SendCoin.fee.multiply(BigDecimal("4"))), MinterSDK.DEFAULT_COIN))
//        ))
//        Thread.sleep(1000 * 3.toLong())
//        MinterExplorerApi.getInstance().getGateApiService().removeHttpInterceptor(gateMock)
    }

    @Test
    fun testSendCoinsSuccessfully() {
//        selectTab(1)
//        waitForBalance(100.0)
//        waitForBalanceUpdate()
//        val sd: SecretData = SecretStorage.generateAddress()
//        val address: String = sd.getMinterAddress().toString()
//
//        // recipient
//        val recipient = Espresso.onView(ViewMatchers.withId(R.id.input_recipient))
//        recipient.perform(ViewActions.replaceText(address))
//        recipient.check(ViewAssertions.matches(ViewMatchers.withText(address)))
//
//        // amount
//        val amount = Espresso.onView(ViewMatchers.withId(R.id.input_amount))
//        amount.perform(ViewActions.replaceText("1"))
//
//        // submit
//        val submit = Espresso.onView(CoreMatchers.allOf(ViewMatchers.withId(R.id.action), ViewMatchers.withText(R.string.btn_send)))
//        submit.check(ViewAssertions.matches(ViewMatchers.isEnabled()))
//        submit.perform(ViewActions.click())
//        val confirmTitle = Espresso.onView(CoreMatchers.allOf(ViewMatchers.isDescendantOfA(ViewMatchers.withId(android.R.id.content)), ViewMatchers.withId(R.id.title)))
//        confirmTitle.check(ViewAssertions.matches(ViewMatchers.withText("You're sending")))
//        Espresso.onView(withId(R.id.dialog_amount)).check(ViewAssertions.matches(ViewMatchers.withText(String.format("%s %s", bdHuman(1.0), MinterSDK.DEFAULT_COIN))))
//        Espresso.onView(withId(R.id.tx_recipient_name)).check(ViewAssertions.matches(ViewMatchers.withText(address)))
//        val confirm = Espresso.onView(CoreMatchers.allOf(ViewMatchers.withId(R.id.action_confirm), ViewMatchers.withText(R.string.btn_send)))
//        confirm.perform(ViewActions.click())
//        Espresso.onView(ViewMatchers.withId(R.id.tx_description)).check(ViewAssertions.matches(ViewMatchers.withText(R.string.tx_send_success_dialog_description)))
//        Espresso.onView(withId(R.id.tx_recipient_name)).check(ViewAssertions.matches(ViewMatchers.withText(address)))
//        Espresso.onView(ViewMatchers.withId(R.id.action_view_tx))
//                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
//                .check(ViewAssertions.matches(ViewMatchers.withText("VIEW TRANSACTION")))
//
//        // close dialog to prevent leaking
//        Espresso.onView(CoreMatchers.allOf(ViewMatchers.withText("Close"), ViewMatchers.withId(R.id.action_close))).perform(ViewActions.click())
//
//        // 100 - 1 - 0.01 fee  - 98.99
//        waitForBalance(BigDecimal(99).subtract(OperationType.SendCoin.fee))
    }

    @Test
    fun testInputs() {
        selectTab(1)
//        waitForBalance(100.0)
//        val balanceString = String.format("%s (%s)", MinterSDK.DEFAULT_COIN, bdHuman(BigDecimal(100.0)))
//        val submit = Espresso.onView(CoreMatchers.allOf(ViewMatchers.withId(R.id.action), ViewMatchers.withText(R.string.btn_send)))
//        submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        val accountInput = Espresso.onView(ViewMatchers.withId(R.id.input_coin))
//        accountInput.check(ViewAssertions.matches(ViewMatchers.withSubstring(MinterSDK.DEFAULT_COIN)))
//        accountInput.check(ViewAssertions.matches(ViewMatchers.withText(balanceString)))
//        accountInput.perform(ViewActions.click())
//        val accountDialogTitle = Espresso.onView(ViewMatchers.withText(R.string.title_select_account))
//        accountDialogTitle.check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
//        val accountItem = Espresso.onView(CoreMatchers.allOf(ViewMatchers.withId(R.id.item_title), ViewMatchers.withText(balanceString)))
//        accountItem.perform(ViewActions.click())
//
//        // check submit still not enabled
//        submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        val recipient = Espresso.onView(ViewMatchers.withId(R.id.input_recipient))
//        recipient.check(ViewAssertions.matches(Matchers.not(ViewMatchers.withText("aaa"))))
//        recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutHint(R.string.tx_send_recipient_hint)))
//        // set something strange
//        run {
//            recipient.perform(ViewActions.replaceText("aaa"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set something illegal
//        run {
//            recipient.perform(ViewActions.replaceText("whywhywhye"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set too short username, minimum 5 chars
//        run {
//            recipient.perform(ViewActions.replaceText("@aabb"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set too long username, max 16 chars
//        run {
//            recipient.perform(ViewActions.replaceText("@abcdefghijklmnopqrtstuwxyz"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set invalid email #1
//        run {
//            recipient.perform(ViewActions.replaceText("test@fly"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set invalid email #2
//        run {
//            recipient.perform(ViewActions.replaceText("test.fly.com"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set invalid email #2 (but this will parsed as invalid username)
//        run {
//            recipient.perform(ViewActions.replaceText("@gmail.com"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Incorrect recipient format")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//        // set valid username, minimum 5 chars
//        run {
//            recipient.perform(ViewActions.replaceText("@aabbc"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError(null)))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(ViewMatchers.isEnabled()))
//        }
//
//        // set valid username, minimum 5 chars
//        run {
//            recipient.perform(ViewActions.replaceText("@12345"))
//            recipient.check(ViewAssertions.matches(MyMatchers.withInputLayoutError(null)))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(ViewMatchers.isEnabled()))
//        }
//        val amount = Espresso.onView(ViewMatchers.withId(R.id.input_amount))
//        amount.check(ViewAssertions.matches(ViewMatchers.withText("")))
//        amount.check(ViewAssertions.matches(MyMatchers.withInputLayoutHint(R.string.label_amount)))
//
//        // set 0
//        run {
//            amount.perform(ViewActions.replaceText("0"))
//            amount.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Amount must be greater than 0")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set .
//        run {
//            amount.perform(ViewActions.replaceText("."))
//            amount.check(ViewAssertions.matches(MyMatchers.withInputLayoutError("Amount must be greater than 0")))
//            // check submit still not enabled
//            submit.check(ViewAssertions.matches(Matchers.not(ViewMatchers.isEnabled())))
//        }
//
//        // set .1
//        run {
//            amount.perform(ViewActions.replaceText(".1"))
//            amount.check(ViewAssertions.matches(MyMatchers.withInputLayoutError(null)))
//            // check submit is enabled
//            submit.check(ViewAssertions.matches(ViewMatchers.isEnabled()))
//        }
//
//        // maximum is 10^-18
//        run {
//            amount.perform(ViewActions.replaceText("0.102030405060708090"))
//            amount.check(ViewAssertions.matches(MyMatchers.withInputLayoutError(null)))
//        }
//        val useMaxButton = Espresso.onView(withId(R.id.action_maximum))
//        useMaxButton.check(ViewAssertions.matches(ViewMatchers.withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)))
//        useMaxButton.perform(ViewActions.click())
//        amount.check(ViewAssertions.matches(ViewMatchers.withText("100")))
//        val feeLabel = Espresso.onView(ViewMatchers.withId(R.id.fee_label))
//        feeLabel.check(ViewAssertions.matches(ViewMatchers.withText(R.string.tx_send_fee_hint)))
//        val feeText = Espresso.onView(ViewMatchers.withId(R.id.fee_value))
//        feeText.check(ViewAssertions.matches(ViewMatchers.withText(String.format("%s %s", bdHuman(OperationType.SendCoin.fee), MinterSDK.DEFAULT_COIN))))
    }
}